File: AlphabetSorting.c
/*
	Author: Dung Harry
	Date created: 24th, October 2013
	Operating System: Ubuntu 13.04 Desktop x86_64
	Compiler: GNU compiler 4.7.3

	Description: this is the program about sorting an char* array in alphabet order in C programming language
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LENGTH 100

void vEnterNames(char*** lppNames, int numNames);
void vSwapNames(char** lppNames, int iFirstIndex, int iSecondIndex);
int iCharAt(char** lppNames, int numNames, int iNameIndex, int iCharIndex);
void vSortRange(char** lppNames, int numNames, int iIndex, int iHead, int iTail);
int iMaxLength(char** lppNames, int numNames);
void vSortAlphabet(char** lppNames, int numNames);
void vPrint(char** lppNames, int numNames);

void main(int argc, char* argv[]) {
	char **lppNames;
	int iNumNames;

	printf("Author: Dung Harry. Date created: 24th, October 2013\n\n");

	printf("Enter number of names you wanted to sort: ");
	scanf("%d", &iNumNames);

	vEnterNames(&lppNames, iNumNames);

	printf("The list of names you entered is: \n\n");
	vPrint(lppNames, iNumNames);

	printf("\nSorting follows alphabet order...\n");

	vSortAlphabet(lppNames, iNumNames);

	printf("The list of names after sorted is: \n");
	vPrint(lppNames, iNumNames);
}

void vEnterNames(char*** lppNames, int numNames) {
	int i, j;

	*lppNames = (char **) malloc(numNames * sizeof(char *));

	printf("Entering name part...\n\n");

	for(i = 0; i < numNames; i ++) {
		*(*lppNames + i) = (char *) malloc(MAX_LENGTH * sizeof(char));

		printf("Enter name %d: ", i + 1);
		scanf("%s", *(*lppNames + i));

		for(j = 0; (j < MAX_LENGTH - 1) && (*(*(*lppNames + i) + j) != '\0'); j ++);

		if(j == MAX_LENGTH - 2) {
			printf("Your name is too long. The maximum length of name is just 100\n");
		} else {
			for(; ((*(*(*lppNames + i) + j) = getc(stdin)) != '\n') && (j < MAX_LENGTH - 2); j ++);

			if(j == MAX_LENGTH - 2) {
				printf("Your name is too long. The maximum length of name is just 100\n");
			}
		}

		*(*(*lppNames + i) + j) = '\0';
	}

	printf("Finish entering names...\n\n");
}

void vSwapNames(char** lppNames, int iFirstIndex, int iSecondIndex) {
	char* temp = (char *) malloc(MAX_LENGTH * sizeof(char));

	strcpy(temp, *(lppNames + iFirstIndex));
	strcpy(*(lppNames + iFirstIndex), *(lppNames + iSecondIndex));
	strcpy(*(lppNames + iSecondIndex), temp);

	free(temp);
}

int iCharAt(char** lppNames, int numNames, int iNameIndex, int iCharIndex) {
	int iLength;
	int iResult;

	if(iNameIndex < numNames) { 
		iLength = strlen(*(lppNames + iNameIndex));

		if(iCharIndex < iLength) {
			iResult = (int) *(*(lppNames + iNameIndex) + iCharIndex);
		} else {
			iResult = 1;
		}
	} else {
		iResult = 1;
	}

	return iResult;
}

void vSortRange(char** lppNames, int numNames, int iIndex, int iHead, int iTail) {
	int i, j, k, iFirstSum, iSecondSum;

	for(i = iHead; i <= iTail; i ++) {
		for(j = i; j <= iTail; j ++) {
			if(iCharAt(lppNames, numNames, i, iIndex) > iCharAt(lppNames, numNames, j, iIndex)) {
				iFirstSum = iSecondSum = 0;

				for(k = 0; k < iIndex; k ++) {
					iFirstSum += iCharAt(lppNames, numNames, i, k);
					iSecondSum += iCharAt(lppNames, numNames, j, k);
				}

				if(iFirstSum == iSecondSum) {
					vSwapNames(lppNames, i, j);
				}
			}
		}
	}
}

int iMaxLength(char** lppNames, int numNames) {
	int i, iResult = 0;

	for(i = 0; i < numNames; i ++) {
		if(strlen(*(lppNames + i)) > iResult) {
			iResult = strlen(*(lppNames + i));
		}
	}

	return iResult;
}

void vSortAlphabet(char** lppNames, int numNames) {
	int iMax = iMaxLength(lppNames, numNames);
	int iHead, iTail, iTemp, i, j;

	vSortRange(lppNames, numNames, 0, 0, numNames - 1);

	for(j = 1; j <= iMax; j ++) {
		i = 0;
	
		while(i < numNames) {
			iHead = i;
		
			iTemp = iCharAt(lppNames, numNames, i, j - 1);
		
			while(iCharAt(lppNames, numNames, ++ i, j - 1) == iTemp) {
				if(i >= numNames - 1) {
					i ++;
			
					break;
				}
			}
		
			iTail = i - 1;

			vSortRange(lppNames, numNames, j, iHead, iTail);
		}
	}
}

void vPrint(char** lppNames, int numNames) {
	int i;

	for(i = 0; i < numNames; i ++) {
		printf("%d. %s\n", i + 1, *(lppNames + i));
	}
}

File: AlphabetSorting.java
/*
 * 	Author: Dung Harry
 * 	Date created: 29th, September 2013
 * 	JDK 1.7 update 15
 * 	Operating System: Windows 7 Professional x64
 * 	
 * */

import java.util.Scanner;

public class AlphabetSorting {
	private String name[] = new String[100];
	private int numberName = 0;
	
	public AlphabetSorting() {
		this.initialize();
	}
	
	private void initialize() {
		for(int i = 0; i < this.name.length; i ++) {
			this.name[i] = new String("");
		}
	}
	
	public void enterNames() {
		Scanner scanner = new Scanner(System.in);
		String temp = new String("");
		boolean flag = true;
		
		do {
			System.out.print("Enter number of the name you wanted to sort: ");
			temp = scanner.nextLine();
			
			for(int i = 0; i < temp.length(); i ++) {
				if((temp.charAt(i) <= '0') && (temp.charAt(i) >= '9')) {
					flag = false;
				} 
			}
			
			if(!flag) {
				System.out.println("**Error: Your entering is invalid. Enter again");
			} else {
				this.numberName = Integer.parseInt(temp);
				
				if(this.numberName > 100) {
					System.out.println("**Error: Your number of names is greater than 100");
				}
			}
		} while((!flag) || (this.numberName > 100));
		
		System.out.println("Begin entering each name...");
		
		for(int i = 0; i < numberName; i ++) {
			System.out.print("Enter name " + (i + 1) + ": ");
			this.name[i] = scanner.nextLine();
		}
		
		System.out.println("Finish entering name part...");
	}
	
	private void swapNames(int firstIndex, int secondIndex) {
		String temp = new String("");
		
		temp = this.name[firstIndex];
		this.name[firstIndex] = this.name[secondIndex];
		this.name[secondIndex] = temp;
	}
	
	private int charAt(int nameIndex, int index) {
		if((index >= this.name[nameIndex].length()) || (nameIndex >= this.numberName)) {
			return 1;
		} else {
			return this.name[nameIndex].charAt(index);
		}
	}
	
	private void sortRange(int index, int head, int tail) {
		for(int i = head; i <= tail; i ++) {
			for(int j = i; j <= tail; j ++) {
				if(this.charAt(i, index) > this.charAt(j, index)) {
					int firstSum = 0, secondSum = 0;
					
					for(int m = 0; m < index; m ++) {
						firstSum += this.charAt(i, m);
						secondSum += this.charAt(j, m);
					}
					
					if(firstSum == secondSum) {
						this.swapNames(i, j);
					}
				}
			}
		}
	}
	
	private int getMaxLengthName() {
		int tmp = this.name[0].length();
		
		for(int i = 0; i < this.numberName; i ++) {
			if(this.name[i].length() > tmp) {
				tmp = this.name[i].length();
			}
		}
		
		return tmp;
	}
	
	public void sortAlphabet() {
		int maxLength = this.getMaxLengthName();
		int head, tail;
		char temp;
		
		this.sortRange(0, 0, this.numberName - 1);
		
		for(int j = 1; j < maxLength; j ++) {
			int i = 0;
			
			while(i < this.numberName) {
				head = i;
				
				temp = (char) this.charAt(i, j - 1);
				
				while(this.charAt(++ i, j - 1) == temp) {
					if(i >= this.numberName - 1) {
						i ++;
					
						break;
					}
				}
				
				tail = i - 1;
				
				this.sortRange(j, head, tail);
			}
		}
	}
	
	public void printNames() {
		for(int i = 0; i < this.numberName; i ++) {
			System.out.println("Name " + (i + 1) + ": " + this.name[i]);
		}
	}
	
	public static void main(String[] args) {
		AlphabetSorting sorter = new AlphabetSorting();
		sorter.enterNames();
		
		sorter.sortAlphabet();
		
		System.out.println("The result after sorting is: \n");
		
		sorter.printNames();
	}
}
